pkg/operator/connectivitycheckcontroller/connectivity_check_controller.go

```diff
@@ -160,41 +178,66 @@ func (c *connectivityCheckController) Sync(ctx context.Context, syncContext fact
 		return nil
 	}
 
-	checks, err := c.podNetworkConnectivityCheckFn(ctx, syncContext)
+	// Retrieve existing podnetworkconnectivitychecks which is later used for
+	// removing stale objects from k8s.
+	var existingChecks []*v1alpha1.PodNetworkConnectivityCheck
+	if c.checkLister != nil {
+		existingChecks, err = c.checkLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+	}
+
+	newChecks, err := c.podNetworkConnectivityCheckFn(ctx, syncContext)
 	if err != nil {
 		return err
 	}
 
 	pnccClient := c.operatorcontrolplaneClient.ControlplaneV1alpha1().PodNetworkConnectivityChecks(c.namespace)
-	for _, check := range checks {
-		existing, err := pnccClient.Get(ctx, check.Name, metav1.GetOptions{})
+	newCheckNames := sets.NewString()
+	for _, newCheck := range newChecks {
+		newCheckNames.Insert(newCheck.Name)
+		existing, err := pnccClient.Get(ctx, newCheck.Name, metav1.GetOptions{})
 		if err == nil {
-			if equality.Semantic.DeepEqual(existing.Spec, check.Spec) {
+			if equality.Semantic.DeepEqual(existing.Spec, newCheck.Spec) {
 				// already exists, no changes, skip
 				continue
 			}
 			updated := existing.DeepCopy()
-			updated.Spec = *check.Spec.DeepCopy()
+			updated.Spec = *newCheck.Spec.DeepCopy()
 			_, err := pnccClient.Update(ctx, updated, metav1.UpdateOptions{})
```

Comment:
It's the same controller, I think we should use `Apply` here, an example: https://github.com/kubernetes/kubernetes/pull/110173/files